src/librustc_typeck/check/expr.rs

```diff
@@ -566,43 +566,48 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             // the `enclosing_loops` field and let's coerce the
             // type of `expr_opt` into what is expected.
             let mut enclosing_breakables = self.enclosing_breakables.borrow_mut();
-            let ctxt = enclosing_breakables.find_breakable(target_id);
-            if let Some(ref mut coerce) = ctxt.coerce {
-                if let Some(ref e) = expr_opt {
-                    coerce.coerce(self, &cause, e, e_ty);
+            if let Some(ref mut ctxt) = enclosing_breakables.opt_find_breakable(target_id) {
+                if let Some(ref mut coerce) = ctxt.coerce {
+                    if let Some(ref e) = expr_opt {
+                        coerce.coerce(self, &cause, e, e_ty);
+                    } else {
+                        assert!(e_ty.is_unit());
+                        let ty = coerce.expected_ty();
+                        coerce.coerce_forced_unit(self, &cause, &mut |err| {
```

Comment:
Could we refactor this closure into a method or some other parts? It's getting hard to reason about the typesystem logic here cause it's not all in scope at once. :)

Reply:
I would prefer to make that change in a follow up PR to minimize backporting woes.

Reply:
Gonna hold you to that... ;)